iroh/src/downloader.rs

```diff
@@ -0,0 +1,979 @@
+use tokio::sync::{mpsc, oneshot};
+use tokio_util::{sync::CancellationToken, time::delay_queue};
+use tracing::{debug, trace};
+
+mod get;
+mod test;
+
+/// Delay added to a request when it's first received.
+const INITIAL_REQUEST_DELAY: std::time::Duration = std::time::Duration::from_millis(500);
+/// Number of retries initially assigned to a request.
+const INITIAL_RETRY_COUNT: u8 = 4;
+/// Duration for which we keep peers connected after they were last useful to us.
+const IDLE_PEER_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
+/// Capacity of the channel used to comunicate between the [`Downloader`] and the [`Service`].
+const SERVICE_CHANNEL_CAPACITY: usize = 120;
+
+/// Download identifier.
+// Mainly for readability.
+pub type Id = u64;
+
+/// Trait modeling a dialer. This allows for IO-less testing.
+pub trait Dialer:
+    futures::Stream<Item = (PublicKey, anyhow::Result<Self::Connection>)> + Unpin
+{
+    /// Type of connections returned by the Dialer.
+    type Connection: Clone;
+    /// Dial a peer.
+    fn queue_dial(&mut self, peer_id: PublicKey);
+    /// Get the number of dialing peers.
+    fn pending_count(&self) -> usize;
+    /// Check if a peer is being dialed.
+    fn is_pending(&self, peer: &PublicKey) -> bool;
+}
+
+/// Signals what should be done with the request when it fails.
+#[derive(Debug)]
+pub enum FailureAction {
+    /// An error ocurred that prevents the request from being retried at all.
+    AbortRequest(anyhow::Error),
+    /// An error occurred that suggests the peer should not be used in general.
+    DropPeer(anyhow::Error),
+    /// An error occurred in which neither the peer nor the request are at fault.
+    RetryLater(anyhow::Error),
+}
+
+/// Future of a get request.
+type GetFut = LocalBoxFuture<'static, Result<(), FailureAction>>;
+
+/// Trait modelling performing a single request over a connection. This allows for IO-less testing.
+pub trait Getter {
+    /// Type of connections the Getter requires to perform a download.
+    type Connection;
+    /// Return a future that performs the download using the given connection.
+    fn get(&mut self, kind: DownloadKind, conn: Self::Connection) -> GetFut;
+}
+
+/// Concurrency limits for the [`Downloader`].
+#[derive(Debug)]
+pub struct ConcurrencyLimits {
+    /// Maximum number of requests the service performs concurrently.
+    pub max_concurrent_requests: usize,
+    /// Maximum number of requests performed by a single peer concurrently.
+    pub max_concurrent_requests_per_peer: usize,
+    /// Maximum number of open connections the service maintains.
+    pub max_open_connections: usize,
+}
+
+impl Default for ConcurrencyLimits {
+    fn default() -> Self {
+        // these numbers should be checked against a running node and might depend on platform
+        ConcurrencyLimits {
+            max_concurrent_requests: 50,
+            max_concurrent_requests_per_peer: 4,
+            max_open_connections: 100,
```

Comment:
this seems quite high, we might want to default to sth lower here, especially on mobile

Reply:
what do you suggest? I have no idea what the overhead of new connections is in general to pick a reasonable number, maybe we can start with something like 30?

Reply:
(also to note, maybe useful for @b5) that we can set different defaults per platform

Reply:
let's start with 25 (the split of diva's 30 & dig's 20 😄, but also 5 docs worth of gossip connections)

Reply:
done in aba53f2e